engines/sword1/sword1.cpp

```diff
@@ -655,64 +727,72 @@ uint8 SwordEngine::mainLoop() {
-			if (newTime - frameTime < 1000 / FRAME_RATE) {
-				scrollFrameShown = _screen->showScrollFrame();
-				delay((1000 / (FRAME_RATE * 2)) - (_system->getMillis() - frameTime));
+			if (!Logic::_scriptVars[NEW_PALETTE]) {
+				newTime = _system->getMillis();
+				if ((int32)(newTime - frameTime) < _targetFrameTime / 2) {
+					scrollFrameShown = _screen->showScrollFrame();
+					pollInput((_targetFrameTime / 2) - (_system->getMillis() - frameTime));
+				}
 			}
 
+			_sound->engine();
+
 			newTime = _system->getMillis();
-			if ((newTime - frameTime < 1000 / FRAME_RATE) || (!scrollFrameShown))
+			if (((int32)(newTime - frameTime) < _targetFrameTime) || (!scrollFrameShown))
 				_screen->updateScreen();
-			delay((1000 / FRAME_RATE) - (_system->getMillis() - frameTime));
+			pollInput((_targetFrameTime) - (_system->getMillis() - frameTime));
+
+			_vblCount = 0; // Reset the vBlank counter for the other timers...
 
 			_mouse->engine(_mouseCoord.x, _mouseCoord.y, _mouseState);
 
-			if (_systemVars.forceRestart)
-				retCode = CONTROL_RESTART_GAME;
-
-			// The control panel is triggered by F5 or ESC.
-			else if (((_keyPressed.keycode == Common::KEYCODE_F5 || _keyPressed.keycode == Common::KEYCODE_ESCAPE)
-			          && (Logic::_scriptVars[MOUSE_STATUS] & 1)) || (_systemVars.controlPanelMode)) {
-				retCode = _control->runPanel();
-				if (retCode == CONTROL_NOTHING_DONE)
-					_screen->fullRefresh();
-			}
+			retCode = checkKeys();
 
 			_mouseState = 0;
 			_keyPressed.reset();
+
 		} while ((Logic::_scriptVars[SCREEN] == Logic::_scriptVars[NEW_SCREEN]) && (retCode == 0) && (!shouldQuit()));
 
 		if ((retCode == 0) && (Logic::_scriptVars[SCREEN] != 53) && _systemVars.wantFade && (!shouldQuit())) {
-			_screen->fadeDownPalette();
-			int32 relDelay = (int32)_system->getMillis();
-			while (_screen->stillFading()) {
-				relDelay += (1000 / FRAME_RATE);
-				_screen->updateScreen();
-				delay(relDelay - (int32)_system->getMillis());
-			}
+			_screen->startFadePaletteDown(1);
 		}
 
-		_sound->quitScreen();
-		_screen->quitScreen(); // close graphic resources
-		_objectMan->closeSection(Logic::_scriptVars[SCREEN]); // close the section that PLAYER has just left, if it's empty now
+		_screen->quitScreen(); // Close graphic resources
+		waitForFade();
+
+		_sound->quitScreen(); // Purge the sound AFTER they've been faded
+
+		_objectMan->closeSection(Logic::_scriptVars[SCREEN]); // Close the section that PLAYER has just left, if it's empty now
 	}
 	return retCode;
 }
 
-void SwordEngine::delay(int32 amount) { //copied and mutilated from sky.cpp
+void SwordEngine::waitForFade() {
+	uint32 startTime = _system->getMillis();
+	while (_screen->stillFading()) { // This indirectly also waits for FX to be faded
+		if (_vblCount >= _rate)
+			_vblCount = 0;
+
+		pollInput(0);
+
+		// In the remote event that this wait cycle gets
+		// stuck during debugging, trigger a timeout
+		if (_system->getMillis() - startTime > 2000)
```

Comment:
Is 2 seconds enough? What is the maximum fade duration?

Reply:
Yes; basically the fade happens at a 60Hz rate within the simulated interrupt. Every fade operates for 64 steps, so this means: `16.67 ms/step * 64 steps = 1066.88 ms`. When performing some debugging operation, the timer thread can surely have some hiccups, so I took it to 2 seconds just to be safe.